lib/pengine/group.c

```diff
@@ -20,6 +20,46 @@
 #define VARIANT_GROUP 1
 #include "./variant.h"
 
+static bool
+skip_child_rsc(pe_resource_t *rsc, pe_resource_t *child, GList *only_rsc,
+               unsigned int show_opts)
+{
+    bool star_list = pcmk__list_of_1(only_rsc) &&
+                     pcmk__str_eq("*", g_list_first(only_rsc)->data, pcmk__str_none);
+    bool child_filtered = child->fns->is_filtered(child, only_rsc, FALSE);
+
+    /* If the resource is in only_rsc by name (so, ignoring "*") then allow
+     * it regardless if if it's active or not.
+     */
+    if (!star_list && !child_filtered) {
+        return false;
+
+    } else if (!child_filtered) {
+        /* Allow it if the resource is active, or if we were asked to display
+         * inactive resources.
+         */
+        if (child->fns->active(child, FALSE) ||
+            pcmk_is_set(show_opts, pcmk_show_inactive_rscs)) {
+            return false;
+        }
+
+    /* Check if the parent was asked for. */
+    } else {
+        gboolean check_parent = pcmk__str_in_list(only_rsc, rsc_printable_id(rsc), pcmk__str_none) ||
+                                (strstr(rsc->id, ":") != NULL && pcmk__str_in_list(only_rsc, rsc->id, pcmk__str_none));
+
+        /* But only return the child resource if it is either active or
+         * we are displaying inactive resources.
+         */
+        if (check_parent && (child->fns->active(child, FALSE) ||
+                             pcmk_is_set(show_opts, pcmk_show_inactive_rscs))) {
```

Comment:
Previously if check_parent were true, we would call ->is_filtered() with that, which would call the parent's ->is_filtered(). We've lost that here

Reply:
Hm, can you suggest a test case that would show the problem?  All the test case changes look good to me, which could still mean there's changes that should be made that haven't yet.

Reply:
The previous version called

    child_rsc->fns->is_filtered(child_rsc, only_rsc, print_everything)

whereas the new code only calls is_filtered() with FALSE. A new call with FALSE is fine, it lets us eliminate some common cases, but if check_parent is true, we still need to call is_filtered() with that. This here is for the group children, so we know it's always pe__native_is_filtered(), which with check_parent==TRUE calls the uber_parent()'s is_filtered(..., FALSE). pe__group_is_filtered() will in turn call is_filtered(..., FALSE) for each of its children.

So, a test case should be where the child being checked has is_filtered(..., FALSE)==TRUE, check_parent==TRUE, and is_filtered(uber_parent())==FALSE. It might require a cloned group to hit all that.

Reply:
We do already have test cases that involve cloned groups - it's the last stuff in test_crm_mon before we use crm_mon-partial.xml (search for "in cloned group").

Looking at the test output...

* "Complete test output, filtered by clone name in cloned group" - Looks good.
* "XML output, filtered by clone name in cloned group" - Doesn't match text output.  Either we need to add --show-detail, or if that happens by default for XML, it's not getting propagated.
* "Complete text output, filtered by group name in cloned group" - Looks good.
* "XML output, filtered by group name in cloned group" - Looks good, shows stopped clones.
* "Complete text output, filtered by exact group instance name in cloned group" - Looks good.
* "XML output, filtered by exact group instance name in cloned group" - Looks good.
* "Complete text output, filtered by primitive name in cloned group" - Looks good.
* "XML output, filtered by primitive name in cloned group" - Looks good, shows stopped clones.
* "Complete text output, filtered by exact primitive instance name in cloned group" - Looks good.
* "XML output, filtered by exact primitive instance name in cloned group" - Looks good.

We do not have any test cases where we filter on the name of a stopped clone.  Maybe that'd be worth adding and would show additional problems.  I'd love to have my review here double checked.

Reply:
Looking at it again, the original version had dead code. In pe__group_default():

    if (rsc->fns->is_filtered(rsc, only_rsc, TRUE)) {
        return rc;
    }
    print_everything = pcmk__str_in_list(only_rsc, rsc_printable_id(rsc), pcmk__str_none) ||
                       (strstr(rsc->id, ":") != NULL && pcmk__str_in_list(only_rsc, rsc->id, pcmk__str_none));

We know rsc is a group here, meaning rsc->fns->is_filtered() is pe__group_is_filtered(), which has this:

    } else if (pcmk__str_in_list(only_rsc, rsc_printable_id(rsc), pcmk__str_none)) {
        passes = TRUE;
    } else if (strstr(rsc->id, ":") != NULL && pcmk__str_in_list(only_rsc, rsc->id, pcmk__str_none)) {
        passes = TRUE;

That means print_everything in pe__group_default() can never be true, because ->is_filtered() would have already triggered for those conditions.

That raises the question of what print_everything was trying to do. If it simply was redundant, your new approach is fine.

Reply:
>     * "XML output, filtered by clone name in cloned group" - Resources section does not match text output, which seems off to me.

You're right, the clone should contain the group (similar to the first XML output for mysql-clone-group in .exp, although looking at that I'm wondering whether mysql-proxy should also have the instance number)

Reply:
I think print_everything and pe__group_is_filtered were written at different points to solve different corner cases, but ended up growing into being redundant.  I think the idea behind print_everything was that if we were given the name of the group, or a single instance of a cloned group, in --resource= filtering, we would set print_everything to true so we'd know to print all the group's children.

Reply:
And both miss "XML output, filtered by clone name in cloned group" :)

It sounds like print_everything can always be false for groups. It looks like clones could drop the first half of the print_everything test.

BTW I notice an inefficiency in pe__native_is_filtered(): uber_parent() for a bare primitive (not part of a group or clone) is the primitive itself, so we'll unnecessarily make a recursive call in that case.

Reply:
> And both miss "XML output, filtered by clone name in cloned group" :)

Yeah, I think this is actually due to not having the only_rsc list set up correctly.  My one question here is should XML output require --show-detail to see this stuff, or should that be the default?

> It sounds like print_everything can always be false for groups. It looks like clones could drop the first half of the print_everything test.

I'll remove those.

> BTW I notice an inefficiency in pe__native_is_filtered(): uber_parent() for a bare primitive (not part of a group or clone) is the primitive itself, so we'll unnecessarily make a recursive call in that case.

Easy enough to fix.

Reply:
> > It sounds like print_everything can always be false for groups. It looks like clones could drop the first half of the print_everything test.
> 
> I'll remove those.

This stuff is so much more fragile than I'd like.  Making either of those changes results in test case output changing in ways that it shouldn't.

Reply:
> > And both miss "XML output, filtered by clone name in cloned group" :)
> 
> Yeah, I think this is actually due to not having the only_rsc list set up correctly. My one question here is should XML output require --show-detail to see this stuff, or should that be the default?

iirc XML always behaves as if --show-detail is on

> > > It sounds like print_everything can always be false for groups. It looks like clones could drop the first half of the print_everything test.
> > 
> > 
> > I'll remove those.
> 
> This stuff is so much more fragile than I'd like. Making either of those changes results in test case output changing in ways that it shouldn't.

bizarre, I didn't see a way they could matter

Reply:
> It sounds like print_everything can always be false for groups. It looks like clones could drop the first half of the print_everything test.

To be clear, I was talking about the original code; in the new code, i'd expect check_parent is always false in skip_child_rsc()

Reply:
There's gotta be something else going on here that we are not understanding.  If check_parent is always false, nothing in that last else block of skip_child_rsc can ever be run.  If I just remove that code, the test output also changes.

I've pushed the current state of everything.  I guess we can decide whether it's worth trying to track down these problems or going with what we've got.

Reply:
Doh, I just noticed

    return !passes;

in pe__group_is_filtered(). So it's the *opposite* of print_everything/check_parent ...

What also confused me is that in the new code, check_parent *is* the parent check, rather than whether the parent should be checked. (That doesn't change per child, so the original code was more efficient evaluating that once.)

I wish we had done the inverse of is_filtered(), i.e. should_be_displayed(), to avoid having to follow all these double negatives ...

That raises my original concern again, that we are no longer calling child_rsc->fns->is_filtered(..., TRUE) which we would before if print_everything==TRUE. I think I finally have a test case, in a later comment ...

Reply:
> What also confused me is that in the new code, check_parent _is_ the parent check, rather than whether the parent should be checked. (That doesn't change per child, so the original code was more efficient evaluating that once.)

I've renamed "check_parent" to "parent_passes" and have it passed in to skip_child_rsc as an argument now so it'll only be evaluated the once.

> I wish we had done the inverse of is_filtered(), i.e. should_be_displayed(), to avoid having to follow all these double negatives ...

At the time, I think we were really only focused on calling it like "if (rsc->fns->is_filtered()) { continue; }", which does read pretty well.  I started using the "return !passes" style to help a little bit with the double negatives.  The tests can be written a little bit more straightforwardly and then we only have to negate everything in one place, and hopefully the name of the variable helps readability.